### PR TITLE
update current attributes 

### PR DESCRIPTION
**Pull request by kofifus**
This PR includes a few fixes and additions to Firepad attribute changes:
1. Toolbar button state now changes according to the relevant attributes. Included buttons are bold,Italics, Underline & strike. This can be expanded in the future to include all buttons.
2. The original code did not properly handle attribute changes at the end of the line, for example turning bold off and then inserting a newline would still make text on the new line bold. 
   The reason was the original code backs up before the newline which holds this attribute change. The new code backs up to the correct pos.
3. The original code did not properly handle cursor at the beginning of a non empty line - it would copy the attribute s from the left (line before) and not the right (first char). The new code handles this situation correctly.
4. The original code did not properly handle backspace. The new code preserves the attribute of the removed character when pressing backspace.

Demo at: http://plnkr.co/edit/vGtggEc3xQrurST0z6Cl?p=preview

### CODE REVIEW

File: lib/rich-text-codemirror.js [hunk 1]
```diff
@@ -943,8 +965,39 @@ firepad.RichTextCodeMirror = (function () {
...
+      toggleClass(toolbar.boldBtnElem, 'firepad-btn-highlight', attr.b === true);
+      toggleClass(toolbar.italicBtnElem, 'firepad-btn-highlight', attr.i === true);
+      toggleClass(toolbar.underlineBtnElem, 'firepad-btn-highlight', attr.u === true);
+      toggleClass(toolbar.strikeBtnElem, 'firepad-btn-highlight', attr.s === true);
+    }, 200);
+  }
+
+  // If newState is provided add/remove theClass accordingly, otherwise toggle theClass
+  function toggleClass(elem, theClass, newState) {
+    let matchRegExp = new RegExp('(?:^|\\s)' + theClass + '(?!\\S)', 'g');
```
Comment (jbrecht) [inline lib/rich-text-codemirror.js]: Unless we start building with babelify, you can't use `let` here, it breaks the build.

File: lib/rich-text-codemirror.js [hunk 2]
```diff
@@ -943,8 +965,39 @@ firepad.RichTextCodeMirror = (function () {
...
     }
+
+    this.updateToolbarButnsState_();
   };
 
+  RichTextCodeMirror.prototype.updateToolbarButnsState_ = function() {
+    var toolbar=this.codeMirror.firepad.toolbar;
+    var attr=this.currentAttributes_;
+
+    if (!toolbar.boldBtnElem) { // cache button elements
```
Comment (jbrecht) [inline lib/rich-text-codemirror.js]: unit tests error out here because toolbar is undefined

**Comment by jbrecht** [comment]
awesome, this is super close now! there are two issues that you may or may not wish to address:
1. if you *hold down* the delete button (instead of pressing once per deleted character) and delete all the text in a line, you lose the style. You may have to look into this to solve the problem:
https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent#Auto-repeat_handling

2. while the correct style is applied, for some reason the cursor looks wrong when you delete all the characters. Try this: set the size to something large, like 42, then type some characters. Now delete the characters. You see the cursor shrinks back down until you type a character, then it grows to the 42 size.

Regardless, this PR is a great improvement! If it were my call, I would merge it and treat these two issues I bring up as separate issues to be addressed later.

**Comment by mikelehen** [comment]
Rather than look at raw keys, we should modify the deleteLeft() method in
RichTextCodeMirror (
https://github.com/firebase/firepad/blob/master/lib/rich-text-codemirror.js#L1012
)

**Comment by kofifus** [comment]
Michael I don't think that is possible.. the issue is that after you hit backspace you get the following sequence:
- deleteLeft
- onCursorActivity_ that calls
- updateCurrentAttributes_

so the call to updateCurrentAttributes_ is not issued from deleteLEft and so cannot be prevented there, it needs to be prevented in onCursorActivity and for that we need to know that the last keydown was backspace.

Unless there's another way I'm missing ?

**Comment by kofifus** [comment]
Michael - 

I think I see what you mean and updated the PR (and plunkr). deleteLeft now sets a flag doNotUpdateCurrentAttributesOnNextCursorActivity to tell onNextCursorActivity not to call updateCurrentAttributes. All is well and no need for keyDown handler.

John - 

I could not reproduce your two items above. Can you please retest with new plunkr (http://plnkr.co/edit/mWASMKBMF5w2GP9m9ny7?p=preview) and if still there explain how to reproduce

thx!

**Comment by jbrecht** [comment]
Yes, both issues are reproducible. Steps to reproduce:
**Issue 1**
1. Set font size to 42
2. Type some characters
3. Hold down delete until all characters are deleted
4. Result: Font size reverts to 14. (Also applies to any other style you may apply)

**Issue 2** can be repeated in two ways
A. 
1. Set font size to 42. 
2. Result: cursor does not change until you type the first character.

B. 
1. Set font size to 42
2. Type some characters
3. Delete all the characters by individual delete key presses
4. Type more characters
5. Result: Cursor shrinks back down to size 14 after step 3, then grows back to 42 upon the first key press of step 4

**Comment by kofifus** [comment]
John, I think the issue is not with holding down delete of cursor size but with what happens when you delete all characters in the document. At the moment once you backspace over everything and there's nothing left, if you press backspace again it will reset all attributes. You can see this if you press backspace repeatedly. I am not sure if this is a bad effect or not.

**Comment by jbrecht** [comment]
It's true that issue #1 is only repeatable on the first line of a document, but issue #2 can happen at any line of the document. No matter what line you are in the document, and regardless of what the font size is set to, if the cursor is at the beginning of that line and the line is empty, the cursor size is 14, not the current size.

#1 only occurs when you hold down the delete key. If you press the delete key once for each character you delete, you retain your current attribute even when the document becomes empty. This means, in theory, you can have an empty document and still have attributes. This is no surprise if looked at another way - supposed you start at an empty document, and set a font size or color, the next character you type will be in that font size and color, not the default.

**Comment by kofifus** [comment]
ok, #1 was easy to fix, simply in deleteLeft, if we are at position 0 just exit without doing anything as there's nothing to delete.

please check updated code/plunkr (http://plnkr.co/edit/4dM9dQxma3tJPKQuwISQ?p=preview)

as for #2 I'm not sure where the cursor height is set, is it in FP or CM or the textarea itself ... perhaps Michael can add some info on that.

**Comment by kofifus** [comment]
Hey John, I think no1 is now solved, see new code and plunkr http://plnkr.co/edit/xkoQFcJUwcmNtBBJCtsA?p=info

Would be great if you can give it a good testing !
Thx

**Comment by kofifus** [comment]
Also (as you can see from the zillion commits) I can't get the Travis build to approve my code, all it says is "ERROR [launcher]: PhantomJS failed 2 times (timeout). Giving up." .. is there a way to get more info on what bothers it ?

The code is valid, passes jshint and works well ...

**Comment by mikelehen** [comment]
That means the tests are failing.  You should be able to reproduce locally
with "npm run travis" or "grunt test"  You can also open them in a browser
by opening firepad/test/index.html

-Michael

**Comment by jbrecht** [comment]
Strange. This branch builds fine and passes all tests for me. Unfortunately, it doesn't seem to fix the issue of cursor size or of holding down delete.

**Comment by mikelehen** [comment]
Yeah.  Seems like maybe travis is broken for some other reason. :-/

**Comment by jbrecht** [comment]
@kofifus  - 
looks like package.json has a pretty old version of the phantom launcher:
"karma-phantomjs-launcher": "~0.1.0",

maybe try updating that to current (1.0.0) and see what happens?

**Comment by kofifus** [comment]
John as far as I can see issue 1 is now solved, see http://plnkr.co/edit/MaxY6gGXNgn07fWn3NoW

**Comment by kofifus** [comment]
ok updated package.json - all good :)

**Comment by vincentwoo** [comment]
i pegged travis to node 5 for now which should stop the test bleeding, too

**Comment by kofifus** [comment]
Vincent - as you can see from the package.json I committed, phantomjs has been renamed to phantomjs-prebuilt. Can you pls update FP package.js accordingly ?

John - can you please verify your issue 1 above is solved with new code ?

Thx!

**Comment by jbrecht** [comment]
I still see issue #1 unfortunately.

**Comment by kofifus** [comment]
I can't see how John, following your steps above and plunkr http://plnkr.co/edit/vGtggEc3xQrurST0z6Cl?p=preview

1. Set font size to 42
2. Type some characters
3. Hold down delete until all characters are deleted
4. Type some characters

Result: step 4 characters are in font 42

Am I missing something ?

**Comment by jbrecht** [comment]
Oh, that is very strange. Yes, it works correctly in your Plunker-deployed version, but not in my local one. There must be some other problem with my local version, I'll see if I can figure out what it is.

**Comment by jbrecht** [comment]
oh man, I found the answer and it is devious. in the deleteLeft() function, you have this code:
`var charLeft=cm.getRange({line: cursorPos.line, ch: cursorPos.ch-1}, cursorPos);`
`console.log('<'+charLeft+'> '+(charLeft===''));`
`if (charLeft==='') return; `

(What you checked in doesn't have that logging statement, but your plunker does, so I checked it out.) So, in your Plunker, that charLeft does in fact become empty when you get back to the beginning, but in mine, it ends up being HTML entity #8203. I'm not sure exactly why, but apparently it's a known issue:
http://stackoverflow.com/questions/18478847/why-is-8203-being-injected-into-my-html 

So maybe I can come up with a fix that checks if it is empty OR if it is that entity?

**Comment by kofifus** [comment]
I don't understand how you would get char 8203 in your FP instance. I think it would be best to investigate how it gets there first before making a code change around it. As far as I can see firepad does not use/insert this char ?

**Comment by mikelehen** [comment]
8203 is a zero-width space character.  firepad won't insert it
automatically, but you could paste one in or something...

**Comment by jbrecht** [comment]
I still don't understand it. If I inspect the span containing my text everything is fine until I delete the last character and then that entity is inserted. it's *only* there when the content becomes empty. The only other clue is that when there is text, the span looks like so:
`<span class="  firepad-fs-42px">abc</span>`

but when I delete all the characters ('abc'), the span then looks like so:
`<span cm-text="">​</span>`

**Comment by kofifus** [comment]
Just to verify, can you use the same CodeMirror version as in the plunkr - https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.14.2/codemirror.js

**Comment by mikelehen** [comment]
Hrm, strange. I could imagine that maybe CodeMirror inserts &8203; into the
HTML temporarily for some reason, but I would *not* expect getRange() to
reflect that.  Are you able to reproduce this behavior with the checked-in
firepad examples?

The span class disappearing when the text is deleted doesn't surprise me.

**Comment by kofifus** [comment]
Just to verify, can you try using the same CodeMirror version as in the plunkr - https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.14.2/codemirror.js

**Comment by jbrecht** [comment]
I did double check that we have the same code mirror versions. Here is the CodeMirror function that inserts this entity:
`  var zwspSupported;
  function zeroWidthElement(measure) {`
    if (zwspSupported == null) {
      var test = elt("span", "\u200b");
      removeChildrenAndAdd(measure, elt("span", [test, document.createTextNode("x")]));
      if (measure.firstChild.offsetHeight != 0)
        zwspSupported = test.offsetWidth <= 1 && test.offsetHeight > 2 && !(ie && ie_version < 8);
    }
    var node = zwspSupported ? elt("span", "\u200b") :
      elt("span", "\u00a0", null, "display: inline-block; width: 1px; margin-right: -1px");
    node.setAttribute("cm-text", "");
    return node;
  }`

**Comment by jbrecht** [comment]
based on the comment in the place where this function is called, it looks like it's job is purposefully to ensure a line is never actually empty:
     ` // Ensure at least a single node is present, for measuring.`
      `if (builder.map.length == 0)`
    `    builder.map.push(0, 0, builder.content.appendChild(zeroWidthElement(cm.display.measure)));`

I can't explain why this happens for me in my app, but not in Plunker or for you guys in yours. It's something about how zwspSupported is determined, which is possibly influence by CSS that I have in my app but you don't have in yours.

**Comment by mikelehen** [comment]
Regardless, this seems like an internal implementation detail of CodeMirror
and it shouldn't be visible in the result of cm.getRange()...  May want to
ping the CodeMirror folks...

**Comment by kofifus** [comment]
If I open https://codemirror.net/ and delete everything in the CM area there and inspect it I see the <span>​&#8203</span>, but it is inside the CodeMirror-measure div which CM uses internally for measuring. That is all fine. I cannot see how getRange would return it.

I think you should produce a simple sample. an empty FP (or ideally CM) and a button that alerts with the getRange result code and see if you can reproduce it.

**Comment by jbrecht** [comment]
I am positively baffled. This only happens when firepad/codemirror are embedded in my app. if I run a stripped down version on its own, that character does not appear as part of the document. regardless, if i add a little bit of code in my version to catch it, this definitely fixes the delete problem.

**Comment by kofifus** [comment]
ok possibly somewhere in your code you put that char in the editor rather than FP/CM bug. Either way let's remove this discussion from the PR and if you can provide a sample pls post an issue to the FP/CM groups. thx!

**Pull request closed without merging**